File: skills/ruby-on-rails/assets/BridgeButtonComponent.kt
// Android: register in your Application subclass →
//   Hotwire.registerBridgeComponents(BridgeComponentFactory("button", ::ButtonComponent))
import android.util.Log
import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

class ButtonComponent(
    name: String,
    private val delegate: BridgeDelegate<HotwireDestination>
) : BridgeComponent<HotwireDestination>(name, delegate) {

    override fun onReceive(message: Message) {
        when (message.event) {
            "connect" -> handleConnectEvent(message)
            else -> Log.w("ButtonComponent", "Unknown event for message: $message")
        }
    }

    private fun handleConnectEvent(message: Message) {
        val data = message.data<MessageData>() ?: return
        // Add a toolbar menu item on delegate.destination titled data.title
        // whose click calls performButtonClick().
    }

    private fun performButtonClick(): Boolean = replyTo("connect")

    @Serializable
    data class MessageData(@SerialName("title") val title: String)
}

File: skills/ruby-on-rails/assets/BridgeButtonComponent.swift
// iOS: register in AppDelegate → Hotwire.registerBridgeComponents([ButtonComponent.self])
import HotwireNative
import UIKit

final class ButtonComponent: BridgeComponent {
    override class var name: String { "button" }

    override func onReceive(message: Message) {
        guard let viewController else { return }
        addButton(via: message, to: viewController)
    }

    private var viewController: UIViewController? {
        delegate?.destination as? UIViewController
    }

    private func addButton(via message: Message, to viewController: UIViewController) {
        guard let data: MessageData = message.data() else { return }

        let action = UIAction { [unowned self] _ in
            self.reply(to: "connect")
        }
        viewController.navigationItem.rightBarButtonItem = UIBarButtonItem(title: data.title, primaryAction: action)
    }
}

private extension ButtonComponent {
    struct MessageData: Decodable {
        let title: String
    }
}
